src/lib/equinoxe/equinoxe.ml

```diff
@@ -227,3 +227,112 @@ module Make (B : Backend) : API with type 'a io = 'a B.io = struct
       Http.get ~t ~path
   end
 end
+
+module MakeFriendly (B : Backend) : FRIENDLY_API with type 'a io = 'a B.io =
+struct
+  type 'a io = 'a B.io
+
+  module A : API with type 'a io = 'a B.io = Make (B)
+
+  type t = A.t
+
+  let return x = B.return x
+  let ( let* ) m f = B.bind f m
+  let ( let+ ) m f = B.map f m
+  let fail msg = B.fail (`Msg msg)
+
+  let extract_from json field ~filter =
+    match Ezjsonm.find_opt json [ field ] with
+    | Some v -> filter v
+    | None ->
+        let msg =
+          Format.sprintf "Field %s is not available in output json" field
+        in
+        fail msg
+
+  let extract_string_from json field =
+    let filter = function
+      | `String s -> return s
+      | _ ->
+          let msg = Format.sprintf "Field %s is not a string" field in
+          fail msg
+    in
+    extract_from json field ~filter
+
+  let extract_int_from json field =
+    let filter = function
+      | `Float f when Float.is_integer f -> return (int_of_float f)
+      | _ ->
+          let msg = Format.sprintf "Field %s is not an integer" field in
+          fail msg
+    in
+    extract_from json field ~filter
+
+  let extract_array_from json field =
+    let filter = function
+      | `A l -> return l
+      | _ ->
+          let msg = Format.sprintf "Field %s is not an array" field in
+          fail msg
+    in
+    extract_from json field ~filter
+
+  let create = A.create
+
+  module Orga = struct
+    type config = {
+      name : string;
+      id : string;
+      account_id : string;
+      website : string;
+      maintenance_email : string;
+      max_projects : int;
+    }
+
+    let extract_config_from json =
+      let* name = extract_string_from json "name" in
+      let* id = extract_string_from json "id" in
+      let* account_id = extract_string_from json "account_id" in
+      let* website = extract_string_from json "website" in
+      let* maintenance_email = extract_string_from json "maintenance_email" in
+      let* max_projects = extract_int_from json "max_projects" in
+      return { id; account_id; name; website; maintenance_email; max_projects }
+
+    let to_string config =
+      let pp_empty s = if s = "" then "<empty>" else s in
+      Format.sprintf
+        "{\n\
+         \tname: %s;\n\
+         \tid: %s;\n\
+         \taccount_id: %s;\n\
+         \twebsite: %s;\n\
+         \tmaintenance_email: %s;\n\
+         \tmax_projects: %d;\n\
+         }\n"
+        (pp_empty config.id)
+        (pp_empty config.account_id)
+        (pp_empty config.name) (pp_empty config.website)
+        (pp_empty config.maintenance_email)
+        config.max_projects
+
+    let get_all t =
+      let* json = A.Orga.get_organizations t in
+      let* organizations = extract_array_from json "organizations" in
+      List.fold_left
+        (fun acc dict ->
+          let* acc = acc in
+          let+ config = extract_config_from dict in
+          config :: acc)
+        (return []) organizations
```

Comment:
It would probably be easier to use the Ezjsonm lib directly than to redefine the functions `extract_xxx_from`:

```ocaml
let config_of_json j : config =
  { name = Ezjsonm.(find j ["name"] |> get_string) ; ... }

let get_all t =
  let* json = http_blabla t in
  try return @@ Ezjsonm.get_list orga_of_json json
  with Ezjsonm.Parse_error (value, err) ->
    fail (Printf.sprintf "Orga.get_all: parse error %S on %s from %s" err (to_string value) (to_string json))
```
The bonus is that the user gets the name of function that failed. And in fact, this parse error is more indicative of a bug with our library than something the user is expected to fix (so perhaps we should raise something else than \``Msg`?).

I'm not against introducing helper functions where Ezjsonm doesn't help. For example, the `Ezjsonm.find` function raises `Not_found` which doesn't fit with my proposed try/catch:

```ocaml
let get field t =
  try Ezjsonm.find t [field]
  with Not_found ->
    raise Ezjsonm.Parse_error (t, Printf.sprintf "get %S" field)
```

Reply:
You are right, it will simplify the way to track errors if we have to fix some bugs. I need to rely more on `Ezjsonm`, instead of reinventing the wheel! :sweat_smile:

Reply:
For the error we could extend the the type in backend with an `Internal_error of string. Thanks to it, the user would know that it is related to an internal issue with the library.

Reply:
Then the error could be a bit more structured than a string! (so that we don't have to inline "print" the json in the message)